EmptyLoopBodyFixer - introduction

SpacePossum opened the pull request:
```
possum@aquarium:/home/seb/work/PHP-CS-Fixer$ php php-cs-fixer describe empty_loop_body
Xdebug: [Step Debug] Could not connect to debugging client. Tried: localhost:9003 (fallback through xdebug.client_host/xdebug.client_port) :-(
Description of empty_loop_body rule.
Empty loop-body must be in configured style.

Fixer is configurable using following option:
* style ('braces', 'semicolon'): style of empty loop-bodies; defaults to 'braces'

Fixing examples:
 * Example #1. Fixing with the default configuration.
   ---------- begin diff ----------
   --- Original
   +++ New
   @@ -1,1 +1,1 @@
   -<?php while(foo());
   +<?php while(foo()){}
   
   ----------- end diff -----------

 * Example #2. Fixing with configuration: ['style' => 'semicolon'].
   ---------- begin diff ----------
   --- Original
   +++ New
   @@ -1,1 +1,1 @@
   -<?php while(foo()){}
   +<?php while(foo());
   
   ----------- end diff -----------
```

julienfalque commented on tests/Fixer/ControlStructure/EmptyLoopBodyFixerTest.php:
```diff
@@ -0,0 +1,144 @@
+
+echo 1;
+',
+            ['style' => 'semicolon'],
```
Could you add tests with comments like e.g.
```php
while (foo()) {
    // noop
}
```
In this example, the body shouldn't be considered empty IMO.

SpacePossum replied:
why not, if you use braces as style it won't be touched, if you want to use `;` than it will be fixed and leave the comment,
if you want to use `;` but have exceptions as this one than don't use the rule

julienfalque replied:
Fair enough.

julienfalque commented on src/Fixer/ControlStructure/EmptyLoopBodyFixer.php:
```diff
@@ -0,0 +1,146 @@
+    /**
+     * @internal
+     */
+    public const STYLE_BRACES = 'braces';
```
Can we make those constants private?

SpacePossum replied:
than we cannot use these in unit test, I'm fine either way

julienfalque replied:
I would say it's a good thing: if we change the value of a constant, it's a BC break and the test will prevent that.

julienfalque commented on src/Fixer/ControlStructure/EmptyLoopBodyFixer.php:
```diff
@@ -0,0 +1,146 @@
+            (new FixerOptionBuilder('style', 'Style of empty loop-bodies.'))
+                ->setAllowedTypes(['string'])
+                ->setAllowedValues([self::STYLE_BRACES, self::STYLE_SEMICOLON])
+                ->setDefault(self::STYLE_BRACES)
```
I think `;` is better to express that the empty loop is intentional, I would make it the default.

julienfalque replied:
Not enough to assume they want to follow the `braces` style IMO. /cc @nicolas-grekas maybe?

nicolas-grekas replied:
better use `{}` I suppose since it's a bit more readable/explicit? I've no strong opinion on the topic.

julienfalque replied:
Ok, let's keep `braces` as default value then.

keradus replied:
I would love to put one more argument. real example from our own codebase:
```diff
        for (++$index; $tokens[$index]->isWhitespace(" \t") || $tokens[$index]->isComment(); ++$index);
        for (++$index; $tokens[$index]->isWhitespace(" \t") || $tokens[$index]->isComment(); ++$index) {
        }
```

- version using `{}` - oh, we had a loop that was doing sth one day, and now we no longer execute anything inside, so the whole loop can be dropped! awesome, dead code detected, removing ;) ## ups, I broke the code
- version using `;` - ah, we are moving the index, the code is needed.

(so basically what @julienfalque said)

-----

I strongly suggest to use `;` by default.
@nicolas-grekas , can you reconsider your opinion?

@julienfalque , @SpacePossum  - I suggest to make fixer default `;` whatever gonna be decision of for Symfony ruleset

keradus replied:
(ref) https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5745#issuecomment-889763849
> I've changed the default and removed the rule from the Symfony set. This way the talk about (default) value in sets can be done in new PR's with representation of the community of the set, while not blocking this PR from proceeding.

@nicolas-grekas , don't hesitate to create new ticket to update Symfony set

julienfalque commented on src/Fixer/ControlStructure/EmptyLoopBodyFixer.php:
```diff
@@ -0,0 +1,146 @@
+     */
+    public const STYLE_SEMICOLON = 'semicolon';
+
+    private $tokenLoopKinds = [T_FOR, T_FOREACH, T_WHILE];
```
Can be a constant.

SpacePossum replied:
that would be at the cost of visibility, unless you mean something else ( https://3v4l.org/mEY0G )

julienfalque replied:
I meant
```php
private const TOKEN_LOOP_KINDS = [T_FOR, T_FOREACH, T_WHILE];
```

julienfalque reviewed (commented):
:+1:

SpacePossum commented on src/RuleSet/Sets/SymfonySet.php:
```diff
@@ -49,6 +49,7 @@ public function getRules(): array
             'clean_namespace' => true,
             'concat_space' => true,
             'echo_tag_syntax' => true,
+            'empty_loop_body' => true,
```
no, I just checked the repo

SpacePossum replied:
talk continues here https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5745#discussion_r640626708

keradus reviewed (commented):
let's (re-)align on https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5745#discussion_r679564664 before merging

SpacePossum commented:
I've changed the default and removed the rule from the Symfony set. This way the talk about (default) value in sets can be done in new PR's with representation of the community of the set, while not blocking this PR from proceeding.
